src/llvm_interpreter.rs

```diff
@@ -0,0 +1,753 @@
+    println!("@@@@@@@@@@ after adding global definitions, module:\n{:?}", module);
+
+    let box_unbox_functions = BoxUnboxFunctions {
+        box_i64: gen_box_fn_for_type(&mut module, BalloonTypeTag::Integer, int64_type(), cast_i64_to_i64),
+        unbox_i64: gen_unbox_for_type(&mut module, BalloonTypeTag::Integer, int64_type(), cast_i64_to_i64)
+    };
+
+
+    gen_box_fn_for_type(&mut module, BalloonTypeTag::Float, float64_type(), cast_f64_to_i64);
+    gen_box_fn_for_type(&mut module, BalloonTypeTag::Boolean, int1_type(), cast_ilower_to_i64);
+    println!("@@@@@@@@@@ after adding box definitions, module:\n{:?}", module);
+
+    gen_add_fn(&mut module, &box_unbox_functions);
+    println!("@@@@@@@@@@ after adding unbox definitions, module:\n{:?}", module);
+
+    box_unbox_functions
+}
+
+
+
+
+fn compile_literal(mut module: &mut Module, bb: LLVMBasicBlockRef, box_unbox_functions: &BoxUnboxFunctions, literal: &Literal) -> LLVMValueRef {
+    unsafe {
+        match literal {
+            &Literal::Integer(i64val) => {
+                let builder = Builder::new();
+                builder.position_at_end(bb);
+                let i = LLVMConstInt(int64_type(), i64val as c_ulonglong, 1);
+                let box_intp = LLVMBuildCall(builder.builder,
+                                    box_unbox_functions.box_i64,
+                                    [i].as_mut_ptr(),
+                                    1,
+                                    module.new_string_ptr("valp"));
+                box_intp
+            },
+            &Literal::Float(f64val) => LLVMConstReal(float64_type(), f64val),
+            &Literal::Bool(boolval) =>  LLVMConstInt(int1_type(), boolval as c_ulonglong, 1),
+            other => panic!("unknown literal expr: {:?}", other)
+        }
+    }
+}
+fn compile_expr(module: &mut Module, bb: LLVMBasicBlockRef, box_unbox_functions: &BoxUnboxFunctions, expr: &Expr) -> LLVMValueRef {
+    match expr {
+        &Expr::Literal(ref literal) => compile_literal(module, bb, box_unbox_functions, literal),
+        &Expr::BinaryExpression(ref leftexpr, ref op, ref rightexpr) => unsafe {
+            let builder = Builder::new(); builder.position_at_end(bb);
+
+            let leftbox = compile_expr(module, bb, box_unbox_functions, &leftexpr.data);
+            let rightbox = compile_expr(module, bb, box_unbox_functions, &rightexpr.data);
+
+            let leftval = LLVMBuildCall(builder.builder, box_unbox_functions.unbox_i64,
+                                        [leftbox].as_mut_ptr(), 1, module.new_string_ptr("leftval"));
+            let rightval = LLVMBuildCall(builder.builder, box_unbox_functions.unbox_i64,
+                                         [rightbox].as_mut_ptr(), 1, module.new_string_ptr("rightval"));
+
+            let finalval = match op {
+                Add => LLVMBuildNSWAdd(builder.builder, leftval, rightval, module.new_string_ptr("addval")),
```

Comment:
This isn't matching against `BinaryOp::Add` and instead seems to be wildcard matching creating a new binding `Add`, which means none of the other patterns ever get matched.